File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IdentifiedLocation.scala
package org.enso.compiler.core.ir

import org.enso.syntax.text.Location

import java.util.UUID

/** Couples a location with a possible source identifier.
  *
  * @param location the code location.
  * @param id       the identifier for the location.
  */
case class IdentifiedLocation(location: Location, id: Option[UUID]) {

  /** @return the character index of the start of this source location.
    */
  def start: Int = location.start

  /** @return the character index of the end of this source location.
    */
  def end: Int = location.end

  /** @return the length in characters of this location.
    */
  def length: Int = location.length
}

object IdentifiedLocation {

  /** Utility constructor, building a location without an ID.
    *
    * @param location the code location.
    * @return an [[IdentifiedLocation]] corresponding to the input location.
    */
  def apply(location: Location): IdentifiedLocation =
    IdentifiedLocation(location, None)
}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/IdentifiedLocation.java
package org.enso.compiler.core.ir;

import java.util.UUID;
import scala.Option;

public record IdentifiedLocation(int start, int end, UUID uuid) {

  public IdentifiedLocation(Location location, UUID uuid) {
    this(location.start(), location.end(), uuid);
  }

  public IdentifiedLocation(Location location) {
    this(location.start(), location.end(), null);
  }

  /**
   * @return location of this identified location.
   */
  public Location location() {
    return new Location(start, end);
  }

  /**
   * @return the length in characters of this location.
   */
  public int length() {
    return end - start;
  }

  /**
   * @return option with/out UUID
   */
  public Option<UUID> id() {
    return Option.apply(uuid());
  }

  @Override
  public String toString() {
    return "IdentifiedLocation[start=" + start + ", end=" + end + ", uuid=" + id() + "]";
  }
}
